onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h

```diff
@@ -18,49 +18,175 @@
+                                              const double* Aarray[], int lda,
+                                              const double* Barray[], int ldb,
+                                              const double* beta,
+                                              double* Carray[], int ldc,
+                                              int batch_count,
+                                              const cudaDeviceProp& /*prop*/) {
+  return cublasDgemmBatched(handle,
+                            transa,
+                            transb,
+                            m, n, k,
+                            alpha,
+                            Aarray, lda,
+                            Barray, ldb,
+                            beta,
+                            Carray, ldc,
+                            batch_count);
+}
+inline cublasStatus_t cublasGemmBatchedHelper(cublasHandle_t handle,
+                                              cublasOperation_t transa,
+                                              cublasOperation_t transb,
+                                              int m, int n, int k,
+                                              const half* alpha,
+                                              const half* Aarray[], int lda,
+                                              const half* Barray[], int ldb,
+                                              const half* beta,
+                                              half* Carray[], int ldc,
+                                              int batch_count,
+                                              const cudaDeviceProp& prop) {
+  onnxruntime::cuda::CublasMathModeSetter math_mode_setter(prop, handle, CUBLAS_TENSOR_OP_MATH);
+#ifdef ENABLE_TRAINING
+  float h_a = onnxruntime::math::halfToFloat(*reinterpret_cast<const uint16_t*>(alpha));
+  float h_b = onnxruntime::math::halfToFloat(*reinterpret_cast<const uint16_t*>(beta));
+
+  // accumulating in FP32
+  return cublasGemmBatchedEx(handle,
+                             transa,
+                             transb,
+                             m, n, k,
+                             &h_a,
+                             (const void**)Aarray, CUDA_R_16F, lda,
+                             (const void**)Barray, CUDA_R_16F, ldb,
+                             &h_b,
+                             (void**)Carray, CUDA_R_16F, ldc,
+                             batch_count,
+                             CUDA_R_32F,
+                             CUBLAS_GEMM_DEFAULT);
+#else
+  // accumulating in FP16
+  return cublasHgemmBatched(handle,
+                            transa,
+                            transb,
+                            m, n, k,
+                            alpha,
+                            (const __half**)Aarray, lda,
+                            (const __half**)Barray, ldb,
+                            beta,
+                            (__half**)Carray, ldc,
+                            batch_count);
```

Comment:
>batch_count); [](start = 28, length = 13)

nit: Maybe explicitly set CUDA_R_16F here to avoid confusion?

Reply:
Maybe I misunderstand your comment. Were you saying to specify CUDA_R_16F when calling cublasHgemmBatched? It doesn't support to set data type.

Reply:
`CUDA_R_32F` is a case. In another case, we can also explicitly provide that default argument (I wrongly through the default argument is `CUDA_R_16F`, sorry) for clarify.